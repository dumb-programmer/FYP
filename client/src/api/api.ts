export function login(data: { email: string; password: string }) {
  return fetch("http://localhost:3000/login", {
    method: "POST",
    body: JSON.stringify(data),
    headers: {
      "Content-Type": "application/json",
    },
    mode: "cors",
    credentials: "include",
  });
}

export function signup(data: {
  firstName: string;
  lastName: string;
  email: string;
  password: string;
}) {
  return fetch("http://localhost:3000/signup", {
    method: "POST",
    body: JSON.stringify(data),
    headers: {
      "Content-Type": "application/json",
    },
    mode: "cors",
  });
}

export function getMessages(chatId: string, page: number) {
  return fetch(`http://localhost:3000/chats/${chatId}/messages?page=${page}`, {
    mode: "cors",
    credentials: "include",
  });
}

export function searchMessages(chatId: string, query: string, page: number) {
  return fetch(
    `http://localhost:3000/chats/${chatId}/messages?query=${query}&page=${page}`,
    {
      mode: "cors",
      credentials: "include",
    }
  );
}

export function deleteMessage(chatId: string, messageId: string) {
  return fetch(`http://localhost:3000/chats/${chatId}/messages/${messageId}`, {
    method: "DELETE",
    mode: "cors",
    credentials: "include",
  });
}

export function sendPrompt(data: { prompt: string }, chatId?: string) {
  return fetch(`http://localhost:3000/chats/${chatId}`, {
    method: "POST",
    mode: "cors",
    credentials: "include",
    body: JSON.stringify(data),
    headers: {
      "Content-Type": "application/json",
    },
  });
}

export function getChats(page: number) {
  return fetch(`http://localhost:3000/chats?page=${page}`, {
    mode: "cors",
    credentials: "include",
  });
}

export function updateChat(chatId: string, name: string) {
  return fetch(`http://localhost:3000/chats/${chatId}`, {
    method: "PATCH",
    body: JSON.stringify({ name }),
    headers: {
      "Content-Type": "application/json",
    },
    credentials: "include",
    mode: "cors",
  });
}

export function deleteChat(chatId: string) {
  return fetch(`http://localhost:3000/chats/${chatId}`, {
    method: "DELETE",
    credentials: "include",
    mode: "cors",
  });
}
