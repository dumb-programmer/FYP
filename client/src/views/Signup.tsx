import googleIcon from "@/assets/google-icon.svg";
import githubIcon from "@/assets/github-icon.svg";
import { useForm } from "react-hook-form";
import { zodResolver } from "@hookform/resolvers/zod";
import { SignupSchema } from "@/types/schema";
import { Link } from "react-router-dom";
import LoadingIcon from "@/components/LoadingIcon";

export default function Signup() {
    const { register, handleSubmit, setError, formState: { errors, isSubmitting } } = useForm({
        defaultValues: {
            email: "",
            firstName: "",
            lastName: "",
            password: "",
            confirmPassword: ""
        },
        resolver: zodResolver(SignupSchema)
    });

    return <main className="full-screen centered">
        <div style={{ boxShadow: "0 0 3px 3px rgba(0,0,0,0.2)", padding: 20, borderRadius: 5 }}>
            <form className="form" onSubmit={handleSubmit(async (data) => {
                const response = await fetch("http://localhost:3000/signup", {
                    method: "POST", body: JSON.stringify({ ...data, confirmPassword: undefined }), headers: {
                        "Content-Type": "application/json"
                    }
                })
                if (response.status === 409) {
                    setError("email", { message: "A user with this email already exists" })
                }
            })}>
                <h1>Signup</h1>
                <div style={{ display: "flex", gap: "1rem", flexWrap: "wrap" }}>
                    <div className="form-control">
                        <label className="label" htmlFor="firstName">First Name</label>
                        <input className="input" id="firstName" type="text" {...register("firstName")} />
                        <span className="error-msg">{errors.firstName?.message}</span>
                    </div>
                    <div className="form-control">
                        <label className="label" htmlFor="lastName">Last Name</label>
                        <input className="input" id="lastName" type="text" {...register("lastName")} />
                        <span className="error-msg">{errors.lastName?.message}</span>
                    </div>
                </div>
                <div className="form-control">
                    <label className="label" htmlFor="email">Email</label>
                    <input className="input" id="email" {...register("email")} />
                    <span className="error-msg">{errors.email?.message}</span>
                </div>
                <div className="form-control">
                    <label className="label" htmlFor="password">Password</label>
                    <input className="input" id="password" type="password" {...register("password")} />
                    <span className="error-msg">{errors.password?.message}</span>
                </div>
                <div className="form-control">
                    <label className="label" htmlFor="confirmPassword">Confirm Password</label>
                    <input className="input" id="confirmPassword" type="password" {...register("confirmPassword")} />
                    <span className="error-msg">{errors.confirmPassword?.message}</span>
                </div>
                <button className="btn btn-primary">
                    {
                        isSubmitting ? <LoadingIcon size={25} stroke="white" /> : "Submit"
                    }
                </button>
                <p className="subtle-text">Already have an account? <Link className="link" to="/login">Login</Link></p>
                <hr className="separator" />
                <div className="provider-btns">
                    <a aria-label="signup with google" href="http://localhost:3000/auth/google" className="btn btn-secondary">
                        <img src={googleIcon} style={{ transform: "scale(1.3)" }} alt="google" />
                    </a>
                    <a aria-label="signup with github" href="http://localhost:3000/auth/github" className="btn btn-secondary">
                        <img src={githubIcon} style={{ transform: "scale(1.8)" }} alt="github" />
                    </a>
                </div>
            </form>
        </div>
    </main>;
}