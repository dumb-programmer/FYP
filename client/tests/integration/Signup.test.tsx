import React from "react";
import { render, screen } from "@testing-library/react";
import { expect, describe, it, vi } from "vitest";
import Signup from "../../src/views/Signup";
import { MemoryRouter } from "react-router-dom";
import userEvent from "@testing-library/user-event";
import { signup } from "../../src/api/api";

vi.mock("../../src/api/api", async () => ({ ... await vi.importActual("../../src/api/api"), signup: vi.fn(() => ({ status: 201 })) }))

const renderComponent = () => render(
    <MemoryRouter>
        <Signup />
    </MemoryRouter>
);

describe("Signup", () => {
    it("renders correctly", () => {
        const { container } = renderComponent();
        expect(container).toMatchSnapshot();
    });

    it("firstName input works", async () => {
        renderComponent();

        const input = screen.getByLabelText(/^first name$/i);

        await userEvent.type(input, "test");

        expect(input.value).toEqual("test");
    });

    it("lastName input works", async () => {
        renderComponent();

        const input = screen.getByLabelText(/^last name$/i);

        await userEvent.type(input, "test");

        expect(input.value).toEqual("test");
    });

    it("email input works", async () => {
        renderComponent();

        const input = screen.getByLabelText(/^email$/i);

        await userEvent.type(input, "test");

        expect(input.value).toEqual("test");
    });

    it("password input works", async () => {
        renderComponent();

        const input = screen.getByLabelText(/^password$/i);

        await userEvent.type(input, "test");

        expect(input.value).toEqual("test");
    });

    it("confirmPassword input works", async () => {
        renderComponent();

        const input = screen.getByLabelText(/^confirm password$/i);

        await userEvent.type(input, "test");

        expect(input.value).toEqual("test");
    });

    it("firstName is validated", async () => {
        renderComponent();

        const input = screen.getByLabelText(/^first name$/i);

        await userEvent.click(screen.getByRole("button"));

        expect(screen.getByText(/^first name is required/i)).toBeInTheDocument();

        await userEvent.type(input, Array.from({ length: 200 }).map(() => "a").join(""))

        await userEvent.click(screen.getByRole("button"));

        expect(screen.getByText(/^first name must be 100 characters or fewer/i)).toBeInTheDocument();
    });

    it("lastName is validated", async () => {
        renderComponent();

        const input = screen.getByLabelText(/^last name$/i);

        await userEvent.click(screen.getByRole("button"));

        expect(screen.getByText(/^last name is required/i)).toBeInTheDocument();

        await userEvent.type(input, Array.from({ length: 200 }).map(() => "a").join(""))

        await userEvent.click(screen.getByRole("button"));

        expect(screen.getByText(/^last name must be 100 characters or fewer/i)).toBeInTheDocument();
    });

    it("email is validated", async () => {
        renderComponent();

        const input = screen.getByLabelText(/^email$/i);

        await userEvent.click(screen.getByRole("button"));

        expect(screen.getByText(/^email is required/i)).toBeInTheDocument();

        await userEvent.type(input, "test");

        await userEvent.click(screen.getByRole("button"));

        expect(screen.getByText(/^invalid email/i)).toBeInTheDocument();
    });

    it("password is validated", async () => {
        renderComponent();

        const input = screen.getByLabelText(/^password$/i);

        await userEvent.type(input, "test");

        await userEvent.click(screen.getByRole("button"));

        expect(screen.getByText(/^password must be at least 8 characters/i)).toBeInTheDocument();
    });

    it("confirmPassword is validated", async () => {
        renderComponent();

        const confirmPasswordInput = screen.getByLabelText(/^confirm password$/i);

        await userEvent.click(screen.getByRole("button"));

        expect(screen.getByText(/^confirm password must be at least 8 characters/i)).toBeInTheDocument();

        const passwordInput = screen.getByLabelText(/^password$/i);

        await userEvent.type(passwordInput, "test1234");

        await userEvent.type(confirmPasswordInput, "test1235");

        await userEvent.click(screen.getByRole("button"));

        expect(screen.getByText(/^passwords must match/i)).toBeInTheDocument();
    });

    it("signup is called, if all the inputs are valid", async () => {
        renderComponent();

        const FIRST_NAME = "lorem";
        const LAST_NAME = "epsum"
        const EMAIL = "test@gmail.com";
        const PASSWORD = "12345678";

        await userEvent.type(screen.getByLabelText(/^first name$/i), FIRST_NAME);
        await userEvent.type(screen.getByLabelText(/^last name$/i), LAST_NAME);
        await userEvent.type(screen.getByLabelText(/^email$/i), EMAIL);
        await userEvent.type(screen.getByLabelText(/^password$/i), PASSWORD);
        await userEvent.type(screen.getByLabelText(/^confirm password$/i), PASSWORD);

        await userEvent.click(screen.getByRole("button"));

        expect(signup).toHaveBeenCalled();
    });

    it("On status 409 email is already taken error is shown", async () => {
        signup.mockReturnValue({status:409})

        renderComponent();

        const FIRST_NAME = "lorem";
        const LAST_NAME = "epsum"
        const EMAIL = "test@gmail.com";
        const PASSWORD = "12345678";

        await userEvent.type(screen.getByLabelText(/^first name$/i), FIRST_NAME);
        await userEvent.type(screen.getByLabelText(/^last name$/i), LAST_NAME);
        await userEvent.type(screen.getByLabelText(/^email$/i), EMAIL);
        await userEvent.type(screen.getByLabelText(/^password$/i), PASSWORD);
        await userEvent.type(screen.getByLabelText(/^confirm password$/i), PASSWORD);

        await userEvent.click(screen.getByRole("button"));

        expect(signup).toHaveBeenCalled();

        expect(screen.getByText(/^a user with this email already exists/i))
    })
})