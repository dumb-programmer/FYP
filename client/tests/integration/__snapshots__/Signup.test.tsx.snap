// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Signup > renders correctly 1`] = `
<div>
  <main
    class="min-h-screen flex justify-center items-center"
  >
    <div
      class="card shadow-md"
    >
      <form
        class="flex flex-col gap-4 p-10"
      >
        <h1
          class="card-title text-2xl"
        >
          Signup
        </h1>
        <div
          class="flex gap-4 flex-wrap"
        >
          <div
            class="form-control w-full md:w-auto"
          >
            <label
              class="label"
              for="firstName"
            >
              First Name
            </label>
            <input
              class="input input-bordered"
              id="firstName"
              name="firstName"
              type="text"
            />
            <span
              class="text-sm text-red-600"
            />
          </div>
          <div
            class="form-control w-full md:w-auto"
          >
            <label
              class="label"
              for="lastName"
            >
              Last Name
            </label>
            <input
              class="input input-bordered"
              id="lastName"
              name="lastName"
              type="text"
            />
            <span
              class="text-sm text-red-600"
            />
          </div>
        </div>
        <div
          class="form-control"
        >
          <label
            class="label"
            for="email"
          >
            Email
          </label>
          <input
            class="input input-bordered"
            id="email"
            name="email"
          />
          <span
            class="text-sm text-red-600"
          />
        </div>
        <div
          class="form-control"
        >
          <label
            class="label"
            for="password"
          >
            Password
          </label>
          <input
            class="input input-bordered"
            id="password"
            name="password"
            type="password"
          />
          <span
            class="text-sm text-red-600"
          />
        </div>
        <div
          class="form-control"
        >
          <label
            class="label"
            for="confirmPassword"
          >
            Confirm Password
          </label>
          <input
            class="input input-bordered"
            id="confirmPassword"
            name="confirmPassword"
            type="password"
          />
          <span
            class="text-sm text-red-600"
          />
        </div>
        <button
          class="btn btn-primary"
        >
          Submit
        </button>
        <p
          class="text-gray-500"
        >
          Already have an account? 
          <a
            class="link"
            href="/login"
          >
            Login
          </a>
        </p>
        <div
          class="divider"
        >
          OR
        </div>
        <div
          class="flex flex-col gap-4"
          data-testid="auth-provider-btns"
        >
          <a
            class="btn glass flex items-center gap-4"
            href="http://localhost:3000/auth/google"
          >
            <img
              alt="google"
              src="/src/assets/google-icon.svg"
              style="transform: scale(1.3);"
            />
            Continue With Google
          </a>
          <a
            class="btn glass flex items-center gap-4"
            href="http://localhost:3000/auth/github"
          >
            <img
              alt="github"
              src="/src/assets/github-icon.svg"
              style="transform: scale(1.8);"
            />
            Continue With Github
          </a>
        </div>
      </form>
    </div>
  </main>
</div>
`;
